concordium-consensus/src/Concordium/KonsensusV1/Consensus/Quorum.hs

```diff
@@ -0,0 +1,238 @@
+    ) =>
+    -- |The 'QuorumMessage' we are receiving.
+    QuorumMessage ->
+    -- |Result of receiving the 'QuorumMessage'.
+    m ReceiveQuorumMessageResult
+receiveQuorumMessage qm@QuorumMessage{..} = process =<< get
+  where
+    process skovData
+        -- The round of the quorum signature message is obsolete.
+        | qmRound < (skovData ^. roundStatus . rsCurrentRound) =
+            return Rejected
+        -- The epoch of the quorum signature message is obsolete.
+        | qmEpoch <= skovData ^. skovEpochBakers . currentEpoch =
+            return Rejected
+        -- The consensus runner is not caught up.
+        | qmEpoch > skovData ^. skovEpochBakers . currentEpoch =
+            return CatchupRequired
+        | otherwise = do
+            case getFinalizerByIndex skovData of
+                -- Signee is not in the finalization committee so we flag it and stop.
+                Nothing -> do
+                    flag (NotAFinalizer qmFinalizerIndex qm)
+                    return Rejected
+                Just FinalizerInfo{..}
+                    -- Check whether the signature is ok or not.
+                    | not (checkQuorumSignatureSingle (getQuorumSignatureMessage skovData) finalizerBlsKey qmSignature) ->
+                        return Rejected
+                    -- Finalizer already signed a message for this round.
+                    | isDoubleSigning finalizerIndex skovData -> do
+                        flag (DoubleSigning qmFinalizerIndex qm)
+                        return Rejected
+                    -- Continue verifying by looking up the block.
+                    | otherwise -> do
+                        getRecentBlockStatus qmBlock skovData >>= \case
+                            -- The signatory signed an already signed block. We flag and stop.
+                            OldFinalized -> do
+                                flag (SignedInvalidBlock qmFinalizerIndex qmBlock qm)
+                                return Rejected
+                            -- The signatory signed an already signed block. We flag and stop.
+                            RecentBlock (BlockFinalized _) -> do
+                                flag (SignedInvalidBlock qmFinalizerIndex qmBlock qm)
+                                return Rejected
+                            -- The signatory signed a dead block. We flag and stop.
+                            RecentBlock BlockDead -> do
+                                flag (SignedInvalidBlock qmFinalizerIndex qmBlock qm)
+                                return Rejected
+                            -- The block is unknown so catch up.
+                            RecentBlock BlockUnknown ->
+                                return CatchupRequired
+                            -- The block is already pending so wait for it to be executed.
+                            RecentBlock (BlockPending _) ->
+                                return Awaiting
```

Comment:
I am not sure we need to have a distinct `Awaiting` case. We could just use `CatchupRequired`.